libs/vkd3d/swapchain.c

```diff
@@ -1448,10 +1524,65 @@ static bool dxgi_vk_swap_chain_find_compatible_unlocked_present_mode(
     return true;
 }
 
+static bool dxgi_vk_swap_chain_can_use_app_frame_id(struct dxgi_vk_swap_chain *chain)
+{
+    return chain->present.low_latency_state.mode &&
+        chain->request.low_latency_frame_id > chain->present.previous_application_frame_id;
+}
+
+static void dxgi_vk_swap_chain_set_low_latency_state(struct dxgi_vk_swap_chain *chain, struct low_latency_state *low_latency_state)
+{
+    /* It is possible that the Vulkan swapchain does not exist when the application sets
+     * the low latency state. If that is the case, just update the present latency state
+     * and it will be set during dxgi_vk_swap_chain_recreate_swapchain_in_present_task. */
+    if (chain->present.vk_swapchain)
+    {
+        const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+        VkLatencySleepModeInfoNV latency_sleep_mode_info;
+
+        memset(&latency_sleep_mode_info, 0, sizeof(latency_sleep_mode_info));
+        latency_sleep_mode_info.sType = VK_STRUCTURE_TYPE_LATENCY_SLEEP_MODE_INFO_NV;
+        latency_sleep_mode_info.pNext = NULL;
+
+        latency_sleep_mode_info.lowLatencyMode = low_latency_state->mode;
+        latency_sleep_mode_info.lowLatencyBoost = low_latency_state->boost;
+        latency_sleep_mode_info.minimumIntervalUs = low_latency_state->minimum_interval_us;
+
+        VK_CALL(vkSetLatencySleepModeNV(chain->queue->device->vk_device, chain->present.vk_swapchain, &latency_sleep_mode_info));
+    }
+
+    chain->present.low_latency_state = *low_latency_state;
+}
+
+static void dxgi_vk_swap_chain_low_latency_state_update(struct dxgi_vk_swap_chain *chain)
+{
+    if (chain->request.low_latency_update_requested)
+    {
+        /* If the application is changing the low latency mode reset the previously used app frame id. */
+        if (chain->present.low_latency_state.mode != chain->request.requested_low_latency_state.mode)
+            chain->present.previous_application_frame_id = 0;
+
+        pthread_mutex_lock(&chain->present.low_latency_swapchain_lock);
```

Comment:
Is a lock actually needed here since it's only called in present task? Having a lock here implies this can be called externally.

Reply:
I thought it would probably be best to have this lock so calls to the other low latency entry points won't have the present task update the drivers low latency state for their duration.

Reply:
Actually nevermind, the update is atomic within the implementation so this isn't required.